kotlinx-coroutines-core/jvm/test/internal/SegmentQueueTests.kt

```diff
@@ -0,0 +1,211 @@
+package kotlinx.coroutines.internal
+
+import com.devexperts.dxlab.lincheck.LinChecker
+import com.devexperts.dxlab.lincheck.annotations.Operation
+import com.devexperts.dxlab.lincheck.annotations.Param
+import com.devexperts.dxlab.lincheck.paramgen.IntGen
+import com.devexperts.dxlab.lincheck.strategy.stress.StressCTest
+import kotlinx.atomicfu.atomic
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import java.util.*
+import java.util.concurrent.CyclicBarrier
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.concurrent.thread
+import kotlin.random.Random
+import kotlin.test.assertEquals
+
+/**
+ * This queue implementation is based on [SegmentQueue] for testing purposes and is organized as follows. Essentially,
+ * the [SegmentBasedQueue] is represented as an infinite array of segments, each stores one element (see [OneElementSegment]).
+ * Both [enqueue] and [dequeue] operations increment the corresponding global index ([enqIdx] for [enqueue] and
+ * [deqIdx] for [dequeue]) and work with the indexed by this counter cell. Since both operations increment the indices
+ * at first, there could be a race: [enqueue] increments [enqIdx], then [dequeue] checks that the queue is not empty
+ * (that's true) and increments [deqIdx], looking into the corresponding cell after that; however, the cell is empty
+ * because the [enqIdx] operation has not been put its element yet. To make the queue non-blocking, [dequeue] can mark
+ * the cell with [BROKEN] token and retry the operation, [enqueue] at the same time should restart as well; this way,
+ * the queue is obstruction-free.
+ */
+private class SegmentBasedQueue<T>(createFirstSegmentLazily: Boolean) : SegmentQueue<OneElementSegment<T>>(createFirstSegmentLazily) {
```

Comment:
Extract it to a separate file `SegmentBasedQueue.kt` together with its `OneElementSegment` class.